File: backend/src/appearance/svg_geometry.py
"""Bounded absolute geometry for the static SVG profile, before platform drawing."""

from dataclasses import dataclass
import math
import sys

from src.appearance.svg_path import (
    MAX_COMMANDS,
    SvgCommand,
    SvgValidationError,
    parse_path,
)


MAX_COORDINATE = 1_000_000_000_000.0
MAX_EXPANDED_COMMANDS = MAX_COMMANDS * 8


def coordinate(value: float) -> float:
    if not math.isfinite(value) or abs(value) > MAX_COORDINATE:
        raise SvgValidationError("SVG_GEOMETRY_LIMIT")
    return value


def norm(x: float, y: float) -> float:
    scale = max(abs(x), abs(y))
    if scale == 0:
        return 0.0
    a, b = x / scale, y / scale
    return scale * math.sqrt(a * a + b * b)


def ratio_product(value: float, numerator: float, denominator: float) -> float:
    if value == 0:
        return 0.0

    def exponent(number: float) -> int:
        # Match Java Math.getExponent, including its subnormal sentinel.
        return math.frexp(number)[1] - 1 if abs(number) >= sys.float_info.min else -1023

    a, b, c = exponent(value), exponent(numerator), exponent(denominator)
    mantissa = (
        math.ldexp(value, -a) * math.ldexp(numerator, -b) / math.ldexp(denominator, -c)
    )
    try:
        return coordinate(math.ldexp(mantissa, a + b - c))
    except OverflowError as error:
        raise SvgValidationError("SVG_GEOMETRY_LIMIT") from error


def arc(
    x1: float, y1: float, values: tuple[float, ...], x2: float, y2: float
) -> tuple[SvgCommand, ...]:
    if x1 == x2 and y1 == y2:
        return ()
    rx, ry = values[:2]
    if rx == 0 or ry == 0:
        return (SvgCommand("L", (x2, y2)),)
    # fmod matches Kotlin remainder for negative rotations.
    rotation = math.radians(math.fmod(values[2], 360))
    cosine, sine = math.cos(rotation), math.sin(rotation)
    dx, dy = (x1 - x2) / 2, (y1 - y2) / 2
    xp, yp = cosine * dx + sine * dy, -sine * dx + cosine * dy
    corrected_rx = coordinate(norm(xp, ratio_product(yp, rx, ry)))
    corrected_ry = coordinate(norm(yp, ratio_product(xp, ry, rx)))
    if corrected_rx > rx or corrected_ry > ry:
        rx, ry = corrected_rx, corrected_ry
    if rx <= 0 or ry <= 0:
        raise SvgValidationError("SVG_GEOMETRY_LIMIT")
    nx, ny = xp / rx, yp / ry
    distance = norm(nx, ny)
    if distance <= 0 or not math.isfinite(distance):
        raise SvgValidationError("SVG_GEOMETRY_LIMIT")
    sign = -1.0 if values[3] == values[4] else 1.0
    factor = sign * math.sqrt(max(0.0, 1 - distance * distance))
    cxp, cyp = rx * (ny / distance) * factor, -ry * (nx / distance) * factor
    cx = coordinate(cosine * cxp - sine * cyp + (x1 + x2) / 2)
    cy = coordinate(sine * cxp + cosine * cyp + (y1 + y2) / 2)
    ux, uy = (xp - cxp) / rx, (yp - cyp) / ry
    vx, vy = (-xp - cxp) / rx, (-yp - cyp) / ry
    start = math.atan2(uy, ux)
    sweep = math.atan2(ux * vy - uy * vx, ux * vx + uy * vy)
    if values[4] == 0 and sweep > 0:
        sweep -= 2 * math.pi
    if values[4] == 1 and sweep < 0:
        sweep += 2 * math.pi
    if not math.isfinite(sweep):
        raise SvgValidationError("SVG_GEOMETRY_LIMIT")
    count = max(1, math.ceil(abs(sweep) / (math.pi / 4)))
    if count > 8:
        raise SvgValidationError("SVG_GEOMETRY_LIMIT")
    step = sweep / count

    def point(angle: float) -> tuple[float, float]:
        return (
            coordinate(
                cx + rx * cosine * math.cos(angle) - ry * sine * math.sin(angle)
            ),
            coordinate(
                cy + rx * sine * math.cos(angle) + ry * cosine * math.sin(angle)
            ),
        )

    def derivative(angle: float) -> tuple[float, float]:
        return (
            -rx * cosine * math.sin(angle) - ry * sine * math.cos(angle),
            -rx * sine * math.sin(angle) + ry * cosine * math.cos(angle),
        )

    result = []
    for index in range(count):
        a = start + index * step
        b = a + step
        p = (x1, y1) if index == 0 else point(a)
        q = (x2, y2) if index == count - 1 else point(b)
        u, v = derivative(a), derivative(b)
        alpha = 4.0 / 3.0 * math.tan(step / 4)
        result.append(
            SvgCommand(
                "C",
                tuple(
                    coordinate(value)
                    for value in (
                        p[0] + alpha * u[0],
                        p[1] + alpha * u[1],
                        q[0] - alpha * v[0],
                        q[1] - alpha * v[1],
                        q[0],
                        q[1],
                    )
                ),
            )
        )
    return tuple(result)


@dataclass(frozen=True)
class SvgGeometry:
    commands: tuple[SvgCommand, ...]
    length_bound: float
    contours: int
    source_commands: int


def resolve_path(value: str) -> SvgGeometry:
    result: list[SvgCommand] = []
    x = y = sx = sy = length = 0.0
    previous = ""
    cubic: tuple[float, float] | None = None
    quadratic: tuple[float, float] | None = None
    contours = 0

    def append(command: str, values: tuple[float, ...]) -> None:
        nonlocal x, y, sx, sy, length, contours
        if result and result[-1].command == "Z" and command not in {"M", "Z"}:
            # SVG closepath ends the subpath. Drawing afterwards starts another
            # at the same point; make the native contour and dash reset explicit.
            append("M", (x, y))
        for item in values:
            coordinate(item)
        if len(result) >= MAX_EXPANDED_COMMANDS:
            raise SvgValidationError("SVG_GEOMETRY_LIMIT")
        if command == "Z":
            length += norm(x - sx, y - sy)
        elif command != "M":
            px, py = x, y
            for index in range(0, len(values), 2):
                length += norm(values[index] - px, values[index + 1] - py)
                px, py = values[index : index + 2]
        result.append(SvgCommand(command, values))
        if command == "Z":
            x, y = sx, sy
        else:
            x, y = values[-2:]
        if command == "M":
            sx, sy = x, y
            contours += 1

    tokens = parse_path(value)
    for token in tokens:
        code, values, relative = (
            token.command.upper(),
            token.values,
            token.command.islower(),
        )

        def point(index: int) -> tuple[float, float]:
            return (
                coordinate(values[index] + (x if relative else 0)),
                coordinate(values[index + 1] + (y if relative else 0)),
            )

        next_cubic = next_quadratic = None
        if code == "Z":
            append("Z", ())
        elif code in {"M", "L"}:
            append(code, point(0))
        elif code == "H":
            append("L", (values[0] + (x if relative else 0), y))
        elif code == "V":
            append("L", (x, values[0] + (y if relative else 0)))
        elif code == "C":
            a, b, end = point(0), point(2), point(4)
            next_cubic = b
            append("C", a + b + end)
        elif code == "S":
            a = (
                (2 * x - cubic[0], 2 * y - cubic[1])
                if previous in {"C", "S"} and cubic is not None
                else (x, y)
            )
            b, end = point(0), point(2)
            next_cubic = b
            append("C", a + b + end)
        elif code == "Q":
            a, end = point(0), point(2)
            next_quadratic = a
            append("Q", a + end)
        elif code == "T":
            a = (
                (2 * x - quadratic[0], 2 * y - quadratic[1])
                if previous in {"Q", "T"} and quadratic is not None
                else (x, y)
            )
            end = point(0)
            next_quadratic = a
            append("Q", a + end)
        elif code == "A":
            end = point(5)
            for part in arc(x, y, values, *end):
                append(part.command, part.values)
            x, y = end
        else:
            raise SvgValidationError("SVG_PATH_SYNTAX")
        previous, cubic, quadratic = code, next_cubic, next_quadratic
    return SvgGeometry(tuple(result), length, contours, len(tokens))

File: android/app/src/main/java/com/dayforge/domain/appearance/SvgGeometry.kt
package com.dayforge.domain.appearance

import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.ceil
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.sin
import kotlin.math.sqrt
import kotlin.math.tan

internal const val SVG_MAX_COORDINATE = 1_000_000_000_000.0
internal const val SVG_MAX_EXPANDED_COMMANDS = SVG_MAX_COMMANDS * 8

/** Absolute M/L/C/Q/Z geometry; no platform drawing or partial successful parse. */
internal data class SvgGeometry(
    val commands: List<SvgCommand>, val lengthBound: Double, val contours: Int, val sourceCommands: Int
)

private fun coordinate(value: Double): Double {
    svgRequire(value.isFinite() && abs(value) <= SVG_MAX_COORDINATE, "SVG_GEOMETRY_LIMIT")
    return value
}

private fun norm(x: Double, y: Double): Double {
    val scale = max(abs(x), abs(y))
    if (scale == 0.0) return 0.0
    val a = x / scale; val b = y / scale
    return scale * sqrt(a * a + b * b)
}

/** Avoid overflow/underflow in a*b/c before checking the actual derived coordinate. */
private fun ratioProduct(value: Double, numerator: Double, denominator: Double): Double {
    if (value == 0.0) return 0.0
    val a = Math.getExponent(value); val b = Math.getExponent(numerator); val c = Math.getExponent(denominator)
    val mantissa = Math.scalb(value, -a) * Math.scalb(numerator, -b) / Math.scalb(denominator, -c)
    return coordinate(Math.scalb(mantissa, a + b - c))
}

/** SVG endpoint arc conversion, with corrected radii and at most eight cubic pieces. */
private fun arc(x1: Double, y1: Double, values: List<Double>, x2: Double, y2: Double): List<SvgCommand> {
    if (x1 == x2 && y1 == y2) return emptyList()
    var rx = values[0]; var ry = values[1]
    if (rx == 0.0 || ry == 0.0) return listOf(SvgCommand('L', listOf(x2, y2)))
    val rotation = Math.toRadians(values[2] % 360.0)
    val cosine = cos(rotation); val sine = sin(rotation)
    val dx = (x1 - x2) / 2; val dy = (y1 - y2) / 2
    val xp = cosine * dx + sine * dy
    val yp = -sine * dx + cosine * dy
    val correctedRx = coordinate(norm(xp, ratioProduct(yp, rx, ry)))
    val correctedRy = coordinate(norm(yp, ratioProduct(xp, ry, rx)))
    if (correctedRx > rx || correctedRy > ry) { rx = correctedRx; ry = correctedRy }
    svgRequire(rx > 0 && ry > 0, "SVG_GEOMETRY_LIMIT")
    val nx = xp / rx; val ny = yp / ry
    val distance = norm(nx, ny)
    svgRequire(distance > 0 && distance.isFinite(), "SVG_GEOMETRY_LIMIT")
    val sign = if (values[3] == values[4]) -1.0 else 1.0
    val factor = sign * sqrt(max(0.0, 1 - distance * distance))
    val cxp = rx * (ny / distance) * factor
    val cyp = -ry * (nx / distance) * factor
    val cx = coordinate(cosine * cxp - sine * cyp + (x1 + x2) / 2)
    val cy = coordinate(sine * cxp + cosine * cyp + (y1 + y2) / 2)
    val ux = (xp - cxp) / rx; val uy = (yp - cyp) / ry
    val vx = (-xp - cxp) / rx; val vy = (-yp - cyp) / ry
    val start = atan2(uy, ux)
    var sweep = atan2(ux * vy - uy * vx, ux * vx + uy * vy)
    if (values[4] == 0.0 && sweep > 0) sweep -= 2 * Math.PI
    if (values[4] == 1.0 && sweep < 0) sweep += 2 * Math.PI
    svgRequire(sweep.isFinite(), "SVG_GEOMETRY_LIMIT")
    val count = maxOf(1, ceil(abs(sweep) / (Math.PI / 4)).toInt())
    svgRequire(count <= 8, "SVG_GEOMETRY_LIMIT")
    val step = sweep / count
    fun point(angle: Double): Pair<Double, Double> =
        coordinate(cx + rx * cosine * cos(angle) - ry * sine * sin(angle)) to
            coordinate(cy + rx * sine * cos(angle) + ry * cosine * sin(angle))
    fun derivative(angle: Double): Pair<Double, Double> =
        (-rx * cosine * sin(angle) - ry * sine * cos(angle)) to
            (-rx * sine * sin(angle) + ry * cosine * cos(angle))
    return List(count) { index ->
        val a = start + index * step; val b = a + step
        val p = if (index == 0) x1 to y1 else point(a)
        val q = if (index == count - 1) x2 to y2 else point(b)
        val u = derivative(a); val v = derivative(b)
        val alpha = 4.0 / 3.0 * tan(step / 4)
        SvgCommand('C', listOf(p.first + alpha * u.first, p.second + alpha * u.second,
            q.first - alpha * v.first, q.second - alpha * v.second, q.first, q.second).map(::coordinate))
    }
}

internal fun resolveSvgPath(value: String): SvgGeometry {
    val result = mutableListOf<SvgCommand>()
    var x = 0.0; var y = 0.0; var sx = 0.0; var sy = 0.0
    var previous = ' '
    var cubic: Pair<Double, Double>? = null
    var quadratic: Pair<Double, Double>? = null
    var length = 0.0
    var contours = 0
    fun append(command: Char, values: List<Double>) {
        values.forEach(::coordinate)
        svgRequire(result.size < SVG_MAX_EXPANDED_COMMANDS, "SVG_GEOMETRY_LIMIT")
        if (command == 'Z') length += norm(x - sx, y - sy)
        else if (command != 'M') {
            var px = x; var py = y
            values.chunked(2).forEach { pair ->
                length += norm(pair[0] - px, pair[1] - py)
                px = pair[0]; py = pair[1]
            }
        }
        result.add(SvgCommand(command, values))
        if (command == 'Z') { x = sx; y = sy }
        else { x = values[values.size - 2]; y = values.last() }
        if (command == 'M') { sx = x; sy = y; contours++ }
    }
    val tokens = parseSvgPath(value)
    tokens.forEach { token ->
        val code = token.command.uppercaseChar()
        val values = token.values
        val relative = token.command.isLowerCase()
        fun point(index: Int) = coordinate(values[index] + if (relative) x else 0.0) to
            coordinate(values[index + 1] + if (relative) y else 0.0)
        var nextCubic: Pair<Double, Double>? = null
        var nextQuadratic: Pair<Double, Double>? = null
        when (code) {
            'Z' -> append('Z', emptyList())
            'M', 'L' -> point(0).let { append(code, listOf(it.first, it.second)) }
            'H' -> append('L', listOf(values[0] + if (relative) x else 0.0, y))
            'V' -> append('L', listOf(x, values[0] + if (relative) y else 0.0))
            'C' -> {
                val a = point(0); val b = point(2); val end = point(4)
                nextCubic = b
                append('C', listOf(a.first, a.second, b.first, b.second, end.first, end.second))
            }
            'S' -> {
                val control = cubic
                val a = if (previous in "CS" && control != null) (2*x-control.first) to (2*y-control.second) else x to y
                val b = point(0); val end = point(2)
                nextCubic = b
                append('C', listOf(a.first, a.second, b.first, b.second, end.first, end.second))
            }
            'Q' -> {
                val a = point(0); val end = point(2)
                nextQuadratic = a
                append('Q', listOf(a.first, a.second, end.first, end.second))
            }
            'T' -> {
                val control = quadratic
                val a = if (previous in "QT" && control != null) (2*x-control.first) to (2*y-control.second) else x to y
                val end = point(0)
                nextQuadratic = a
                append('Q', listOf(a.first, a.second, end.first, end.second))
            }
            'A' -> {
                val end = point(5)
                arc(x, y, values, end.first, end.second).forEach { append(it.command, it.values) }
                x = end.first; y = end.second
            }
            else -> throw SvgValidationException("SVG_PATH_SYNTAX")
        }
        previous = code
        cubic = nextCubic
        quadratic = nextQuadratic
    }
    return SvgGeometry(result.toList(), length, contours, tokens.size)
}
